corehq/ex-submodules/casexml/apps/case/models.py

```diff
@@ -583,7 +585,13 @@ def _safe_replace_and_force_to_string(me, attr, val):
     
     def update_from_case_update(self, case_update, xformdoc, other_forms=None):
         other_forms = other_forms or {}
-        if case_update.has_referrals():
+        if is_deprecation(xformdoc):
+            # remove the form from actions and xform_ids (based on orig_id)
+            # and rebuild
+            self.xform_ids = [id for id in self.xform_ids if id != xformdoc.orig_id]
+            self.actions = [a for a in self.actions if a.xform_id != xformdoc.orig_id]
```

Comment:
I think there's a design decision here which is to undo the original case action by removing and rebuilding, and then applying the new case action on top of the whole thing, even if there are intervening case actions between the original form submission and the second, deprecating, form submission. Am I reading that right? That seems reasonable—maybe also leave a comment explaining that decision (as opposed to other other plausible decision of swapping in the new case action for the old one in the same position in the list).